WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListViewModel.kt

```diff
@@ -1,13 +1,131 @@
 package org.wordpress.android.ui.prefs.categories
 
-import androidx.lifecycle.ViewModel
+import androidx.annotation.DrawableRes
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import org.wordpress.android.R
+import org.wordpress.android.R.string
+import org.wordpress.android.fluxc.action.TaxonomyAction
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.TaxonomyStore.OnTaxonomyChanged
+import org.wordpress.android.models.CategoryNode
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.ui.posts.GetCategoriesUseCase
+import org.wordpress.android.ui.prefs.categories.CategoriesListViewModel.UiState.Content
+import org.wordpress.android.ui.prefs.categories.CategoriesListViewModel.UiState.Error.GenericError
+import org.wordpress.android.ui.prefs.categories.CategoriesListViewModel.UiState.Error.NoConnection
+import org.wordpress.android.ui.prefs.categories.CategoriesListViewModel.UiState.Loading
+import org.wordpress.android.ui.utils.UiString
+import org.wordpress.android.ui.utils.UiString.UiStringRes
+import org.wordpress.android.util.AppLog
+import org.wordpress.android.util.AppLog.T
+import org.wordpress.android.util.NetworkUtilsWrapper
+import org.wordpress.android.viewmodel.ScopedViewModel
 import javax.inject.Inject
+import javax.inject.Named
 
-class CategoriesListViewModel @Inject constructor() : ViewModel() {
+class CategoriesListViewModel @Inject constructor(
+    private val getCategoriesUseCase: GetCategoriesUseCase,
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher,
+) : ScopedViewModel(bgDispatcher) {
     private var isStarted = false
+    lateinit var siteModel: SiteModel
 
-    fun start() {
+    private val _uiState: MutableLiveData<UiState> = MutableLiveData()
+    val uiState: LiveData<UiState> = _uiState
+
+    fun start(siteModel: SiteModel) {
         if (isStarted) return
         isStarted = true
+
+        this.siteModel = siteModel
+        getCategoriesFromDb()
+        fetchCategoriesFromNetwork(isInvokedFromInit = true)
+    }
+
+    private fun getCategoriesFromDb() {
+        _uiState.postValue(Loading)
+        launch {
+            val siteCategories = getCategoriesUseCase.getSiteCategories(siteModel)
+            if (siteCategories.isNotEmpty())
+                _uiState.postValue(Content(siteCategories))
+        }
+    }
+
+    private fun fetchCategoriesFromNetwork(isInvokedFromInit: Boolean = false) {
+        launch {
+            if (networkUtilsWrapper.isNetworkAvailable()) {
+                if (!isInvokedFromInit)
+                    _uiState.postValue(Loading)
+                getCategoriesUseCase.fetchSiteCategories(siteModel)
+            } else
+                if (_uiState.value is Loading)
+                    _uiState.postValue(NoConnection(::fetchCategoriesFromNetwork))
```

Comment:
@ashiagr Nice catch 🙌 . I believe the UseCase functions are the only ones that need to be called in a background thread . Keeping those inside the launch block, We can move the other statements outside. Also, I will refactor the statements which are updating the UI state in the main thread to use [setValue](https://developer.android.com/reference/android/arch/lifecycle/MutableLiveData#setvalue) instead of PostValue.

Reply:
Have done the necessary changes in https://github.com/wordpress-mobile/WordPress-Android/pull/15552/commits/94d2046ef696f1b1cba318c0c0082dc37a04aab1 . Please have a look 👀